user-interface/src/main/java/life/qbic/datamanager/views/projects/qualityControl/UploadQualityControlDialog.java

```diff
@@ -0,0 +1,281 @@
+
+    // Add upload QualityControls to the link experiment item section, where users can decide on the linked experiment
+    upload.addSucceededListener(this::onUploadSucceeded);
+
+    // Synchronise the Vaadin upload component with the purchase list display
+    // When a file is removed  from the upload component, we also want to remove it properly from memory
+    // and from any additional display
+    upload.getElement().addEventListener("file-remove", this::processClientFileRemoveEvent)
+        .addEventData(VAADIN_FILENAME_EVENT);
+
+    // Put the elements together
+    add(uploadSection, uploadedItemsSection);
+    addClassName("quality-control-upload");
+    confirmButton.setText("Save");
+    // Init the visibility rendering once
+    toggleFileSectionIfEmpty();
+  }
+
+  private void onUploadSucceeded(SucceededEvent succeededEvent) {
+    var qualityControl = new QualityControlItem(succeededEvent.getFileName(),
+        selectableExperimentsForProject);
+    uploadedQualityControlItems.add(qualityControl);
+    qualityControlItemsCache.add(qualityControl);
+    toggleFileSectionIfEmpty();
+  }
+
+  @Override
+  protected void onConfirmClicked(ClickEvent<Button> clickEvent) {
+    fireEvent(new ConfirmEvent(this, true));
+  }
+
+  @Override
+  protected void onCancelClicked(ClickEvent<Button> clickEvent) {
+    fireEvent(new CancelEvent(this, clickEvent.isFromClient()));
+  }
+
+  private void processClientFileRemoveEvent(DomEvent event) {
+    JsonObject jsonObject = event.getEventData();
+    var fileName = jsonObject.getString(VAADIN_FILENAME_EVENT);
+    removeFile(fileName);
+  }
+
+  private void removeFile(String fileName) {
+    removeFileFromBuffer(fileName);
+    removeFileFromDisplay(fileName);
+    toggleFileSectionIfEmpty();
+  }
+
+  private void toggleFileSectionIfEmpty() {
+    boolean filesUploaded = !uploadedQualityControlItems.getChildren().toList().isEmpty();
+    uploadedQualityControlItems.setVisible(filesUploaded);
+    uploadedItemsSectionContent.setVisible(filesUploaded);
+  }
+
+  private void removeFileFromBuffer(String fileName) {
+    multiFileMemoryBuffer.remove(fileName);
+  }
+
+  private void removeFileFromDisplay(String fileName) {
+    qualityControlItemsCache.stream()
+        .filter(qualityControlItem -> qualityControlItem.fileName().equals(
+            fileName)).findAny().ifPresent(qualityControlItem -> {
+          qualityControlItemsCache.remove(qualityControlItem);
+          uploadedQualityControlItems.remove(qualityControlItem);
+        });
+  }
+
+  public List<QualityControlReport> qualityControlItems() {
+    return qualityControlItemsCache.stream().map(this::convertToQualityControl).toList();
+  }
+
+  private QualityControlReport convertToQualityControl(QualityControlItem qualityControlItem) {
+    try {
+      var fileName = qualityControlItem.fileName();
+      var experimentId = qualityControlItem.experimentId();
+      var content = multiFileMemoryBuffer.inputStream(fileName)
```

Comment:
Is this for uploading multiple files at once?

Reply:
Yes i believe it is, the implementation here is inspired by the offer-upload implementation 👍